src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,5392 @@
+	int              fd;
+	uint32_t         num_ents;
+	dfs_obj_t       *dir;
+	long int         offset;
+	struct dfs_mt   *dfs_mt;
+	int              open_flag;
+	int              ref_count;
+	unsigned int     st_ino;
+	daos_anchor_t    anchor;
+	/* path and ents will be allocated together dynamically since they are large. */
+	char            *path;
+	struct dirent   *ents;
+};
+
+struct mmap_obj {
+	/* The base address of this memory block */
+	char            *addr;
+	size_t           length;
+	/* the size of file. It is needed when write back to storage. */
+	size_t           file_size;
+	int              prot;
+	int              flags;
+	/* The fd used when mmap is called */
+	int              fd;
+	/* num_pages = length / page_size */
+	int              num_pages;
+	int              num_dirty_pages;
+	off_t            offset;
+	/* An array to indicate whether a page is updated or not */
+	bool            *updated;
+};
+
+struct fd_dup2 {
+	int fd_src, fd_dest;
+	bool dest_closed;
+};
+
+/* working dir of current process */
+static char            cur_dir[DFS_MAX_PATH] = "";
+static bool            segv_handler_inited;
+/* Old segv handler */
+struct sigaction       old_segv;
+
+/* the flag to indicate whether initlization is finished or not */
+static bool            hook_enabled;
+static pthread_mutex_t lock_dfs;
+static pthread_mutex_t lock_fd;
+static pthread_mutex_t lock_dirfd;
+static pthread_mutex_t lock_mmap;
+static pthread_mutex_t lock_fd_dup2ed;
+
+/* store ! umask to apply on mode when creating file to honor system umask */
+static mode_t          mode_not_umask;
+
+static void
+finalize_dfs(void);
+static void
+update_cwd(void);
+
+struct dir_hdl {
+	d_list_t   entry;
+	dfs_obj_t *oh;
+	char       name[DFS_MAX_PATH];
+};
+
+static inline struct dir_hdl *
+hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dir_hdl, entry);
+}
+
+static bool
+key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return (strncmp(hdl->name, (const char *)key, DFS_MAX_PATH) == 0);
+}
+
+static void
+rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int             rc;
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	rc = dfs_release(hdl->oh);
+	if (rc == ENOMEM)
+		rc = dfs_release(hdl->oh);
+	if (rc && bLog)
+		D_WARN("dfs_release() failed: %s", strerror(rc));
+	D_FREE(hdl);
+}
+
+static bool
+rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	return true;
+}
+
+static uint32_t
+rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return d_hash_string_u32(hdl->name, strnlen(hdl->name, DFS_MAX_PATH));
+}
+
+static d_hash_table_ops_t hdl_hash_ops = {.hop_key_cmp    = key_cmp,
+					  .hop_rec_decref = rec_decref,
+					  .hop_rec_free   = rec_free,
+					  .hop_rec_hash   = rec_hash};
+
+static int
+lookup_insert_dir(int idx_dfs, const char *name, mode_t *mode, dfs_obj_t **obj)
```

Comment:
What this function does is not obvious from the name so I think this would benefit from renaming and a comment.  It looks like it always does a lookup and return the object, inserting it into the hash table if it's a directory?

Reply:
Correct. I copied this function and several related functions from Mohamad's code in ior.
https://github.com/hpc/ior/blob/main/src/aiori-DFS.c#L421
I previously provided reference to the source. He told me to remove it.

ok. I will add comments about this function. Thank you!